packages/fxa-auth-server/lib/payments/iap/apple-app-store/apple-iap.ts

```diff
@@ -0,0 +1,93 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+import { Firestore } from '@google-cloud/firestore';
+import { AppStoreServerAPI, Environment } from 'app-store-server-api';
+import { Container } from 'typedi';
+import { TypedCollectionReference } from 'typesafe-node-firestore';
+
+import { AppConfig, AuthFirestore, AuthLogger } from '../../../types';
+import { IapConfig } from '../types';
+import { PurchaseManager } from './purchase-manager';
+
+export class AppleIAP {
+  private firestore: Firestore;
+  private iapConfigDbRef: TypedCollectionReference<IapConfig>;
+  private log: AuthLogger;
+  private prefix: string;
+
+  public purchaseManager: PurchaseManager;
+  constructor() {
+    const {
+      authFirestore,
+      subscriptions: { appStore },
+    } = Container.get(AppConfig);
+    this.prefix = `${authFirestore.prefix}iap-`;
+    this.firestore = Container.get(AuthFirestore);
+    this.iapConfigDbRef = this.firestore.collection(
+      `${this.prefix}iap-config`
+    ) as TypedCollectionReference<IapConfig>;
+    this.log = Container.get(AuthLogger);
+
+    // Initialize App Store Server API client per bundle ID
+    const appStoreServerApiClients = {};
+    for (const appName of Object.keys(appStore.credentials)) {
+      // Ideally, we'd key by bundleId instead of appName (e.g. 'org.mozilla.ios.FirefoxVPN'
+      // rather than 'guardian-vpn'), but we can't due to
+      // https://github.com/mozilla/node-convict/issues/250, so we will need to map appName
+      // to bundleId.
+      this.getBundleId(appName).then((bundleId) => {
+        AppleIAP.initializeApiClientByBundleId(bundleId, appStore);
+      });
+    }
```

Comment:
This convict workaround isn't pretty.

The challenge is that `getBundleId` is async, while the App Store Server API client is initialized in the `constructor` of the `AppleIAP` class.

Barry had a similar challenge in the `PaymentConfigurationManager` where he [calls the async `load` method in the `constructor`](https://github.com/mozilla/fxa/blob/0f9936330fb31a2cb0cbafe675a46c4ee632598b/packages/fxa-auth-server/lib/payments/configuration/manager.ts#L42), but in his case, his `maybeLoad` helper was only used internally by the `PaymentConfigManager` in known/finite ways. Here, the API client is used by an external module, `PurchaseManager`, and we’d have to remember to add this check for any future method we may add that makes use of the API client.

In a similar vein, I ended up creating a new static, async method on the `AppleIAP` class called `initializeApiClientByBundleId`, and I added a check in the client's caller, `PurchaseManager.querySubscriptionPurchase` to call this method if the client isn’t yet initialized.

Is there a simpler way to do this? Maybe it's okay to s/_/.? That logic was a lot more readable/straightforward.

Reply:
Seems like an antipattern, but I do have yet another idea/option: I thought perhaps if I kept the config keyed off `bundleId`, I could swap out the dots just before `conf.validate()` is called and then swap them back in afterward as a hack.

@bbangert , do you have a preference for any of these three approaches (initialize API client as needed before using, use underscores and string.replace just before creating the API client, or string.replace before/after `conf.validate()`) or any other suggestions?